### PR TITLE
Allow to copy/paste callstacks from Heaptrack to other applications

### PR DESCRIPTION
**Pull request by jeremy-coulon**
copy from Bottom/Up and Top/Down views

very useful for sharing callstacks with other people.

### CODE REVIEW

File: src/analyze/gui/stackstreeview.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#include <QClipboard>
+#include <QKeyEvent>
+
+StacksTreeView::StacksTreeView(QWidget* parent)
+    : QTreeView(parent) {
+
+}
+
+void StacksTreeView::keyPressEvent(QKeyEvent* event) {
+    if(!selectedIndexes().isEmpty() && event->matches(QKeySequence::Copy)) {
```
Comment (milianw) [inline src/analyze/gui/stackstreeview.cpp]: can you instead add a `QShortCut` (probably with `Qt::WidgetWithChildrenShortcut` context) to the tree view and handle its `activated` signal and then fill the clipboard contents from a lambda? That should reduce the boilerplate code required for this feature

File: src/analyze/gui/stackstreeview.hpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#pragma once
```
Comment (milianw) [inline src/analyze/gui/stackstreeview.hpp]: missing copyright header

File: src/analyze/gui/stackstreeview.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#include "stackstreeview.hpp"
```
Comment (milianw) [inline src/analyze/gui/stackstreeview.cpp]: missing copyright header

File: src/analyze/gui/stackstreeview.hpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+#include <QTreeView>
+
+class StacksTreeView : public QTreeView {
+public:
+    StacksTreeView(QWidget* parent = nullptr);
```
Comment (milianw) [inline src/analyze/gui/stackstreeview.hpp]: missing explicit

File: src/analyze/gui/stackstreeview.hpp [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#pragma once
+
+#include <QTreeView>
+
+class StacksTreeView : public QTreeView {
```
Comment (milianw) [inline src/analyze/gui/stackstreeview.hpp]: style: brace on next line

Comment (milianw) [reply src/analyze/gui/stackstreeview.hpp]: also missing Q_OBJECT

File: .gitignore [hunk 6]
```diff
@@ -21,3 +21,6 @@ callgrind.out.*
 *.BASE.*
 *.LOCAL.*
 *.REMOTE.*
+
```
Comment (milianw) [inline .gitignore]: separate commit please

**Comment by jeremy-coulon** [comment]
I took your comments into accounts.
Is it ok for you now?

File: src/analyze/gui/mainwindow.cpp [hunk 7]
```diff
@@ -404,6 +406,26 @@ MainWindow::MainWindow(QWidget* parent)
     m_ui->menu_File->addAction(m_closeAction);
     m_quitAction = KStandardAction::quit(qApp, SLOT(quit()), this);
     m_ui->menu_File->addAction(m_quitAction);
+    QShortcut* shortcut = new QShortcut(QKeySequence(QKeySequence::Copy), m_ui->stacksTree);
+    connect(shortcut, &QShortcut::activated,
```
Comment (milianw) [inline src/analyze/gui/mainwindow.cpp]: always use the four-param connect syntax, i.e. add another context arg (`this` should do)

File: src/analyze/gui/mainwindow.cpp [hunk 8]
```diff
@@ -404,6 +406,26 @@ MainWindow::MainWindow(QWidget* parent)
...
+            [&]()
```
Comment (milianw) [inline src/analyze/gui/mainwindow.cpp]: style: please specify explictly what you capture

File: src/analyze/gui/mainwindow.cpp [hunk 9]
```diff
@@ -404,6 +406,26 @@ MainWindow::MainWindow(QWidget* parent)
...
+    {
+      QTreeView* view = m_ui->stacksTree;
+      if (view->selectionModel()->hasSelection())
```
Comment (milianw) [inline src/analyze/gui/mainwindow.cpp]: style: join next line

File: src/analyze/gui/mainwindow.cpp [hunk 10]
```diff
@@ -404,6 +406,26 @@ MainWindow::MainWindow(QWidget* parent)
...
+      {
+          QString text;
+          QItemSelectionRange range = view->selectionModel()->selection().first();
+          for (auto i = range.top(); i <= range.bottom(); ++i)
```
Comment (milianw) [inline src/analyze/gui/mainwindow.cpp]: dito

File: src/analyze/gui/mainwindow.cpp [hunk 11]
```diff
@@ -404,6 +406,26 @@ MainWindow::MainWindow(QWidget* parent)
...
     m_ui->menu_File->addAction(m_quitAction);
+    QShortcut* shortcut = new QShortcut(QKeySequence(QKeySequence::Copy), m_ui->stacksTree);
+    connect(shortcut, &QShortcut::activated,
+            [&]()
+    {
+      QTreeView* view = m_ui->stacksTree;
+      if (view->selectionModel()->hasSelection())
+      {
+          QString text;
+          QItemSelectionRange range = view->selectionModel()->selection().first();
```
Comment (milianw) [inline src/analyze/gui/mainwindow.cpp]: `const auto range = ...`

though you could also just loop over all ranges and append the texts if there are more than one (disjoint selection)

File: src/analyze/gui/mainwindow.cpp [hunk 12]
```diff
@@ -404,6 +406,26 @@ MainWindow::MainWindow(QWidget* parent)
...
+          for (auto i = range.top(); i <= range.bottom(); ++i)
+          {
+              QStringList rowContents;
+              for (auto j = range.left(); j <= range.right(); ++j)
+                  rowContents << view->model()->index(i,j).data().toString();
+              text += rowContents.join(QString::fromUtf8("\t"));
```
Comment (milianw) [inline src/analyze/gui/mainwindow.cpp]: use QLatin1Char('\t') for join here

File: src/analyze/gui/mainwindow.cpp [hunk 13]
```diff
@@ -404,6 +406,26 @@ MainWindow::MainWindow(QWidget* parent)
...
+              text += QString::fromUtf8("\n");
```
Comment (milianw) [inline src/analyze/gui/mainwindow.cpp]: use QLatin1Char('\n') here

File: src/analyze/gui/mainwindow.cpp (hunk 7 above)
Comment (jeremy-coulon) [reply src/analyze/gui/mainwindow.cpp]: Just tried: 4-param connect syntax doesn't work with lamdba.

File: src/analyze/gui/mainwindow.cpp (hunk 9 above)
Comment (jeremy-coulon) [reply src/analyze/gui/mainwindow.cpp]: I don't know what you mean.

File: src/analyze/gui/mainwindow.cpp (hunk 7 above)
Comment (milianw) [reply src/analyze/gui/mainwindow.cpp]: Why?

http://doc.qt.io/qt-5/qobject.html#connect-5

File: src/analyze/gui/mainwindow.cpp (hunk 9 above)
Comment (milianw) [reply src/analyze/gui/mainwindow.cpp]: make it:

```
if () {
   ...
}
```

instead of

```
if ()
{
    ...
}
```

**Comment by jeremy-coulon** [comment]
I pushed another version including most of your improvements.

File: src/analyze/gui/mainwindow.cpp (hunk 7 above)
Comment (jeremy-coulon) [reply src/analyze/gui/mainwindow.cpp]: ok. my bad!

File: src/analyze/gui/mainwindow.cpp (hunk 9 above)
Comment (jeremy-coulon) [reply src/analyze/gui/mainwindow.cpp]: ok. You could provide a clang-format config file if you want contributors to easily comply with your style.

Comment (milianw) [reply src/analyze/gui/mainwindow.cpp]: very true, I'll have to do that eventually

**Comment by milianw** [comment]
merged, thanks!

**Pull request closed without merging**